src/launcher/java/org/truffleruby/launcher/Launcher.java

```diff
@@ -204,29 +204,42 @@ public static void processArguments(
             List<String> args,
             boolean parseHelpEtc,
             boolean unknownOptionFails,
-            boolean allowedJVMOptions) {
+            boolean isNative) {
 
         try {
             config.setOption(OptionsCatalog.EXECUTION_ACTION, ExecutionAction.UNSET);
 
-            new CommandLineParser(args, config, true, false, parseHelpEtc).processArguments();
+            final CommandLineParser argumentCommandLineParser = new CommandLineParser(args, config, true, false, parseHelpEtc);
+            argumentCommandLineParser.processArguments();
 
             if (unknownOptionFails && !config.getUnknownArguments().isEmpty()) {
                 throw new CommandLineException("unknown option " + config.getUnknownArguments().get(0));
             }
 
             if (config.getOption(OptionsCatalog.READ_RUBYOPT)) {
-                new CommandLineParser(getArgsFromEnvVariable("RUBYOPT"),
-                        config, false, true, true).processArguments();
-                new CommandLineParser(getArgsFromEnvVariable("TRUFFLERUBYOPT"),
-                        config, false, false, true).processArguments();
+                final List<String> rubyoptArgs = getArgsFromEnvVariable("RUBYOPT");
+                final List<String> trufflerubyoptArgs = getArgsFromEnvVariable("TRUFFLERUBYOPT");
+                new CommandLineParser(rubyoptArgs, config, false, true, true).processArguments();
+                new CommandLineParser(trufflerubyoptArgs, config, false, false, true).processArguments();
+
+                if (isNative) {
+                    // Append options from ENV variables to args after last interpreter option, which makes sure that
+                    // maybeExec processes--(native|jvm)* options and filtering them out for new process
+                    // if exec is being called.
+                    // The new process gets all arguments and options including those from ENV variables.
+                    // To avoid processing options from ENV variables twice READ_RUBYOPT option is set to false.
```

Comment:
What about sub-processes then?
They will still use RUBYOPT / TRUFFLERUBYOPT?
And `-X` options are usually not passed down to subprocesses since there is no way to get them at the Ruby level?

Reply:
sub-process should go through the native launcher which will read it once and then `maybeExec`. I think it should be fine. Or do you have a scenario where it would break in mind?

Reply:
It sounds fine.